src/swappy_allocator.rs

```diff
@@ -0,0 +1,442 @@
+// Copyright (c) 2017-present, PingCAP, Inc. Licensed under Apache-2.0.
+
+//! # An [`Allocator`] implementation that has a memory budget and can be
+//! swapped out.
+
+use std::alloc::{AllocError, Allocator, Global, Layout};
```

Comment:
APIs of Allocator have been changed multiple times, this can be hard to maintain until they are stable. Why not fix the problem out side of allocator?

Reply:
I can't think of a better solution without hacking into `VecDeque`, which is quite complex in itself. The allocator, on the other hand, is easy to code and test.

Reply:
I don't get it. For example, set a capacity limit on `VecDeque` and put half of it to disk whenever it's full.

Reply:
I can see it has several drawbacks:
- setting hard limit to individual vecs is not equivalent to a global memory limit.
- maintain a list of items by ourselves that behaves like a `VecDeque` is complex
- involves more overhead than allocator. Because we are adding checks to every container access (in contrary to only adding checks during allocation).

Reply:
> setting hard limit to individual vecs is not equivalent to a global memory limit.

This can also be implemented on a global level limit just like entry cache has done.

> maintain a list of items by ourselves that behaves like a VecDeque is complex

I don't know what items you refer to. To me, they are just backlog pages similar to log entries.

> involves more overhead than allocator. Because we are adding checks to every container access (in contrary to only adding checks during allocation).

The check needs to be examined only when adding new entries. And the cost is smaller than memory copy.

The advantages of doing it without allocator are:
1. It's more flexible and the algorithm can be adjusted to suit best for different cases.
2. We don't rely on unstable APIs, upgrading rustc won't accidentally breaks build.
3. It's clear what has been put into disk, and what's remained in memory, and it should be easier to work with async fetch entries.
4. It's simpler and safer as we don't need to take care about the allocation and deallocation, which is error prone.

Reply:
> They both need to provide VecDeque interfaces

The indexes stored in disk are just like log entries except the content are indexes, they don't need `VecDeque` interface.

> Also, what do you mean "than memory copy"?

What I mean is the check is so simple that its cost is negligible given there are so many copies during writes.

> ...maybe, but that's empty talk until we really need those cases...

Just like what has happen to the memory usage, right?

> I already have working, simple code lying in this PR.

I would argue if it's really as simple as it seems. It requires knowledge of `Allocator` trait, memmap, and ordering to reasoning if it's really correct and safe.

There is another benefit of the doing it out side of allocator, that it should have better and stable performance. As it can only put the old entries to the swap, which is unlikely to be accessed. While using allocator, it behaves the opposite way, that actively access memory is mapped while inactive segment is always kept in memory.

Reply:
First let me say, if needed, a swappable allocator can be used to implement your solution more easily by reusing standard containers.

> `Allocator` trait

Yes. But I would argue the cost of adjusting to new interface is low.

> ordering to reasoning if it's really correct and safe

Maybe you haven't read the code. But the ordering is actually quite simple. All atomics access is relaxed order. Because we use mutex when manipulating pages.

> There is another benefit of the doing it out side of allocator, that it should have better and stable performance.

It's true. We will have to see the performance impact of that. Maybe it will not be a big issue considering the disk write is buffered.

> that actively access memory is mapped while inactive segment is always kept in memory.

I'm not sure. Do you mean inactive segment of the same region? But all entries of the region will be mapped on reallocation.

Reply:
More screenshots are attached. Only read IOs are growing during memory pressure, and it's very likely caused by rewrite process reading out old entries.

Reply:
Why write duration breakdown shows that apply time arise after memory limit is reached?

Reply:
Updates to entry indexes will land on page cache. As mmap doesn't involves syscalls on fast path, I believe the amortized cost is optimal.

Reply:
How about Max Value? I can see both write and apply matches in the end. If max value also matches, it may cause higher jitters.

Reply:
Write duration 999% shows the tail latency is relatively stable.

Reply:
I have checked the monitor, apply can increase as large as writes in tail latency.